### PR TITLE
Avoid Needless Cache Status Fetches in SearchableSnapshotAllocator

### PR DESCRIPTION
**Pull request by original-brownbear**
We shouldn't fetch cache status if no allocation is possible to begin with.
Also, this surfaced an issue with using the `Client` to `reroute` since that
won't retry stale shards (failed the invalid license IT for example) so I moved
to using the `RerouteService` like we do in the `GatewayAllocator`.
(Plus, dried up one method that was 100% the same as in the replica allocator)

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/reroute/ClusterRerouteResponse.java [hunk 1]
```diff
@@ -44,7 +44,7 @@
         explanations = RoutingExplanations.readFrom(in);
     }
 
-    public ClusterRerouteResponse(boolean acknowledged, ClusterState state, RoutingExplanations explanations) {
+    ClusterRerouteResponse(boolean acknowledged, ClusterState state, RoutingExplanations explanations) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/reroute/ClusterRerouteResponse.java:47]: Just a revert of making this public yesterday now that it's not needed for tests any longer.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java [hunk 2]
```diff
@@ -151,20 +156,15 @@ private AllocateUnassignedDecision decideAllocation(RoutingAllocation allocation
...
         final boolean explain = allocation.debugDecision();
-        final MatchingNodes matchingNodes = findMatchingNodes(shardRouting, allocation, fetchedCacheData, explain);
-        assert explain == false || matchingNodes.nodeDecisions != null : "in explain mode, we must have individual node decisions";
-
         // pre-check if it can be allocated to any node that currently exists, so we won't list the cache sizes for it for nothing
         // TODO: in the following logic, we do not account for existing cache size when handling disk space checks, should and can we
         // reliably do this in a world of concurrent cache evictions or are we ok with the cache size just being a best effort hint
         // here?
-        Tuple<Decision, Map<String, NodeAllocationResult>> result = canBeAllocatedToAtLeastOneNode(shardRouting, allocation);
+        Tuple<Decision, Map<String, NodeAllocationResult>> result = ReplicaShardAllocator.canBeAllocatedToAtLeastOneNode(
```
Comment (henningandersen) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java:164]: Can we add a test that we do not trigger any reads or reroutes when deciders say no?

**Review by henningandersen — APPROVED** [review]
LGTM, though it would be good to add a test to verify the problem.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java (hunk 2 above)
Comment (original-brownbear) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java:164]: Sure thing, I pushed 0cb0017 :)

**Comment by original-brownbear** [comment]
Thanks Henning!

**Pull request merged**